src/lib/Block/Block.ts

```diff
@@ -0,0 +1,237 @@
+    FLOW_CWU: 'flow:component-will-unmount',
+    FLOW_RENDER: 'flow:render',
+  };
+
+  public props: TProps;
+
+  protected readonly eventDispatcher = new EventDispatcher();
+
+  protected node?: HTMLElement;
+
+  protected readonly eventBus = new EventBus();
+
+  private _id = v4();
+
+  private readonly _rootElementId: string | undefined;
+
+  private readonly _meta: { tagName: string; props: TProps };
+
+  private _internalEventListeners: Record<string, EventDispatcher> = {};
+
+  private _element!: HTMLElement;
+
+  constructor(
+    tagName = 'div',
+    props: TProps = {} as TProps,
+    _rootElementId?: string,
+  ) {
+    this._meta = {
+      tagName,
+      props,
+    };
+    this.props = this._makePropsProxy({ ...props, __id: this._id });
+    this._rootElementId = _rootElementId;
+    this._registerEvents(this.eventBus);
+
+    this.eventBus.emit(Block.EVENTS.INIT);
+  }
+
+  get element(): HTMLElement {
+    return this._element;
+  }
+
+  leave() {
+    this.eventBus.emit(Block.EVENTS.FLOW_CDU);
+  }
+
+  setProps(nextProps: TProps) {
+    if (!nextProps) {
+      return;
+    }
+    this.eventBus.emit(Block.EVENTS.FLOW_CDU, this.props, nextProps);
+
+    Object.assign(this.props, nextProps);
+  }
+
+  getContent() {
+    return this.element.outerHTML;
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  componentDidMount(): void {}
+
+  // eslint-disable-next-line class-methods-use-this
+  componentWillUnmount(): void {}
+
+  show() {
+    this.eventBus.emit(Block.EVENTS.FLOW_CDU);
+  }
+
+  protected init() {
+    const {
+      tagName,
+      props: { classNames = [] },
+    } = this._meta;
+    const element = document.createElement(tagName);
+    classNames.forEach((name) => {
+      element.classList.add(name);
+    });
+    this._element = element;
+    this.eventDispatcher.node = this._element;
+
+    this.eventBus.emit(Block.EVENTS.FLOW_CDM);
+  }
+
+  /* eslint-disable class-methods-use-this, @typescript-eslint/no-unused-vars */
+
+  // @ts-ignore
+  protected componentDidUpdate(oldProps: TProps, newProps: TProps): boolean {
+    return true;
+  }
+
+  /* eslint-enable class-methods-use-this, @typescript-eslint/no-unused-vars */
+
+  private _registerEvents(eventBus: EventBus) {
+    eventBus.on(Block.EVENTS.INIT, this.init.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDM, this._componentDidMount.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDU, this._componentDidUpdate.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CWU, this._componentWillUnmount.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_RENDER, this._render.bind(this));
+  }
+
+  private _makePropsProxy(props: TProps) {
+    const self = this;
+    return new Proxy(props, {
+      set(target: TProps, prop: string, value) {
+        /* eslint-disable no-param-reassign */
+        // @ts-ignore
+        target[prop] = value;
+        /* eslint-enable no-param-reassign */
+        self.eventBus.emit(Block.EVENTS.FLOW_CDU);
```

Comment:
Можно использовать стрелочные функции, чтобы не прибегать к замыканию
```suggestion
      set: (target: TProps, prop: string, value) => {
        /* eslint-disable no-param-reassign */
        // @ts-ignore
        target[prop] = value;
        /* eslint-enable no-param-reassign */
        this.eventBus.emit(Block.EVENTS.FLOW_CDU);
```

Reply:
Done